### ISSUE TITLE
qv .txt file not loading for sws_8_visit

### ISSUE THREAD

**Issue by HayleyMills** [CLOSER-Cohorts/archivist#223]
I am try to load the sws_8_visit.txt file, but the mappings are not appearing. The in_out_worker is in on, the dataset is linked to the instrument and the .txt file formatting appears OK as far as I can tell. 

[sws_8_visit.txt](https://github.com/CLOSER-Cohorts/archivist/files/1720310/sws_8_visit.txt)

**Comment by charlesdebarros** [comment]
To do: try import QV again after Redis Flush in April

**Comment by charlesdebarros** [comment]
Loading qv.txt the following is available:
qc_9	bsup
qc_8	boccode
qc_17	burine
qc_14	bblood
qc_5	blmpd
qc_6	bcofmed
qc_6	bbptab
qc_3	bnrseid
qc_7_iii	bdepnor
qc_4	bbldaty
qc_4	bbldatm
qc_4	bbldatd
qc_8	bhrtclo
qc_7_i	bhortab
qc_5	blmpy
qc_5	blmpm
qc_6	beptab
qc_6	bstrtab
qc_6	babiot
qc_6	bpkill
qc_7_ii	bpill

From the file attached to this issue, I can see:
qc_3	Bnrseid
qc_4	Bbldatd
qc_4	Bbldatm
qc_4	Bbldaty
qc_5	Blmpd
qc_5	Blmpm
qc_5	Blmpy
qc_6$1;1	Bpkill
qc_6$1;2	Babiot
qc_6$1;3	Bbptab
qc_6$1;4	Bstrtab
qc_6$1;5	Beptab
qc_6$1;6	Bcofmed
qc_7_i	Bhortab
qc_7_i	Bpill
qc_7_ii	Bdepnor
qc_7_iii	Bhrtclo
qc_8	Boccode
qc_9	Bsup
qc_14	Bblood
qc_17	Burine

The entries may not be in the same order but the content is the same.

Would you have a look at this when you have a minute, @HayleyMills?

Thanks.

**Comment by HayleyMills** [comment]
I have checked on the staging server and the mappings which have been loaded and they have not loaded correctly as the mappings are not correct e.g. bhrtclo is mapped to qc_8 instead of qc_7_iii. The topic options are also not displaying. I think this is related to the wider issue of mappings in general.

**Comment by simonreed** [comment]
@HayleyMills the txt file you're importing appears to have the variable name mixed case while the variables in the dataset for that instrument are all in lowercase. That's stopping your new import from being matched, the issue with `bhrtclo is mapped to qc_8 instead of qc_7_iii. ` is because the old `qc_7_iii` mapping was already in the db but your new mapping for `qc_8` is being ignored because of the case sensitive issue. 

I can change it so that the import is not case sensitive, so if forces all variable names to be lowercase when we import them?

To clarify, when you import new mappings for QV do you want to still keep the existing mappings found in the db or do you want to delete the old mappings and only use the ones in a new import?

So in the example of :

Given I view the QV for an instrument from the DB by viewing `/qv.txt` I get:
```
qc_3	bnrseid
qc_7_i	bpilk
qc_7_ii	bdepnox
qc_9	bnrqeid
```

Then I import a `sws_8_visit.txt` file :
```
qc_3	bnrseid
qc_7_i	bpill
qc_7_ii	bdepnor
qc_8	bnrseid
```
Given I view the QV for an instrument from the DB by viewing `/qv.txt`, which option should I see: 

Option A (the exact version from the latest document import, with all old data removed):
```
qc_3	bnrseid
qc_7_i	bpill
qc_7_ii	bdepnor
qc_8	bnrseid
```

Option B (the new mappings are kept but any existing mappings for a question):
```
qc_3	bnrseid
qc_7_i	bpill
qc_7_ii	bdepnor
qc_8	bnrseid
qc_9	bnrqeid
```

Option C (nothing is removed and any new mappings are just amended to the old mappings):
```
qc_3	bnrseid
qc_7_i	bpill
qc_7_i	bpilk
qc_7_ii	bdepnox
qc_7_ii	bdepnor
qc_8	bnrseid
qc_9	bnrqeid
```

**Comment by HayleyMills** [comment]
Thanks Simon, I think if we can remove the case sensitivity that would be really helpful. 

In terms of the options I think either we replace the mappings with the new txt file or we don't do anything as mappings already exist. I think having a mix of both could prove confusing, particularity for large files. Since we do not have any easy way of removing mappings, option A would be the preferred behaviour.

**Comment by simonreed** [comment]
> In terms of the options I think either we replace the mappings with the new txt file or we don't do anything as mappings already exist. I think having a mix of both could prove confusing, particularity for large files. Since we do not have any easy way of removing mappings, option A would be the preferred behaviour.

Just to clarify Option A above does remove any existing mappings and overrides them with just what is included in the txt import file. Therefore it does remove mappings, the only option that doesn't remove any mappings at all is option C (which is what happens now).

**Comment by HayleyMills** [comment]
> Just to clarify Option A above does remove any existing mappings and overrides them with just what is included in the txt import file. Therefore it does remove mappings, the only option that doesn't remove any mappings at all is option C (which is what happens now).

Yes exactly. Thanks for clarification. Option A makes the most sense.

**Comment by simonreed** [comment]
With option C, it would delete all existing mappings before trying to import and match any of the txt file. Which would mean that if no questions/variables were matched then any existing mappings would be deleted and potentially the instrument could be left with no qv mappings. Is that ok @HayleyMills?

**Comment by HayleyMills** [comment]
Yes that's fine for option A. All old mappings removed and all new mappings added (obviously only if they match up to existing variables/questions). If we needed to add new additional mappings we would either have to do these manually in the UI or by exporting the .txt files and then adding the new mappings before re importing.  

It would good to add a warning to say that mappings already exist and they will be deleted, but that's not essential and so it will be fine without.  Warnings and messages generally need improving across Archivist.

**Comment by simonreed** [comment]
> Yes that's fine for option A.

@HayleyMills just to check that was a typo, we're going with option C, right?

**Comment by spuddybike** [comment]
@HayleyMills @simonreed lets go with Option A, where is everything is deleted,  and the contents of the file mappings are implemented. 

In your response, https://github.com/CLOSER-Cohorts/archivist/issues/223#issuecomment-520785417 you say Option C, but that describes Option A, which is where the confusion lies, (i think). So Delete and Replace, NOT update.

For the sake of absolute clarity, an import should remove previous mappings and insert the contents of the new mapping file. 

A quick Google Hangout if its not might be a good idea

**Comment by simonreed** [comment]
@spuddybike you're absolutely right, sorry @HayleyMills you were right it was option A

**Comment by simonreed** [comment]
@HayleyMills this is now on staging

**Comment by HayleyMills** [comment]
@simonreed tried to load a qv.txt mapping file for nshd_89_dsq but its not loading? The message says it was successful.

[nshd_89_dsq.qvmapping.txt](https://github.com/CLOSER-Cohorts/archivist/files/3569102/nshd_89_dsq.qvmapping.txt)

[qv.txt](https://github.com/CLOSER-Cohorts/archivist/files/3569103/qv.txt)

**Comment by simonreed** [comment]
@HayleyMills can you test this again please. The deployment was complete before.

**Comment by HayleyMills** [comment]
Thanks. I've tried to load the qv mappings for nshd_89_dsq and for hcs_2_clinic_11, but they are not loading?

**Comment by HayleyMills** [comment]
Tested in staging again using alspac_01_rat.
The following is all working okay and as expected:
1) The file loads okay.
2) The /qv.txt is correct.
3) If there are incorrect variables/questions in the file it does not load them but logs these in the import log.
4) If there are existing qv mappings it overwrites these.
5) It ignores the case in the question/variable
6) When you load qv file and there are existing qv and tv mappings it flags any conflicts in the UI in red. keeping the existing topics so you can see what they have been assigned to, you can then change these in either the map or dataset view. I think this behaviour is correct.

**Comment by HayleyMills** [comment]
@simonreed Tested in staging, the instrument log page is working okay. 

Outstanding issues: 
a)  I am unable to load qt mappings which are mapped to 0 which is a valid topic of None.  see https://closer-archivist-staging.herokuapp.com/instruments/alspac_99_ttd/imports/24 

b) I also tried to map questions to 0 for the variable which is also returned errors in the log https://closer-archivist-staging.herokuapp.com/instruments/alspac_99_ttd/imports/23 this is correct, as it couldn't find variables of 0 to map to, however, I would expect it not to have overwritten the original mapped qv, since the .txt file failed?

**Comment by HayleyMills** [comment]
This has been resolved in #224
